### PR TITLE
Add Redis backend feature to TFRA

### PR DESCRIPTION
**Pull request by MoFHeKa**
# Description

Brief Description of the PR:

Added Redis as the backend for KV storage and changed the Python interface by calling the Config class and creating/passing the KvCreator class to fit the different backend. Also setting backend parameters with json file.

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Feature

# Checklist:

- [ ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/recommenders-addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running yapf
    - [ ] By running clang-format
- [ ] This PR addresses an already submitted issue for TensorFlow Recommenders-Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

# How Has This Been Tested?

```
redis_config1=tfra.dynamic_embedding.RedisTableConfig(
    redis_config_abs_dir=“/tmp/test/”
)
redis_creator1=tfra.dynamic_embedding.RedisTableCreator(redis_config1)

self.user_embeddings = tfra.dynamic_embedding.get_variable(
            name="user_dynamic_embeddings",
            dim=self.embedding_size,
            initializer=tf.keras.initializers.RandomNormal(-1, 1),
            checkpoint=False,
            KVCreator=redis_creator1
            )
```

or you can do it like this:

```
redis_config_dir = os.path.join(tempfile.mkdtemp(dir=os.environ.get('TEST_TMPDIR')), "save_restore")
redis_config_path = os.path.join(tempfile.mkdtemp(prefix=redis_config_dir), "hash")
os.makedirs(redis_config_path)
redis_config_path = os.path.join(redis_config_path, "redis_config.json")
redis_config_params = {
  "redis_host_ip":["127.0.0.1"],
  "redis_host_port":[6379],
  "using_model_lib":False
}
with open(redis_config_path, 'w', encoding='utf-8') as f:
  f.write(json.dumps(redis_config_params, indent=2, ensure_ascii=True))
redis_config = de.RedisTableConfig(
  redis_config_abs_dir=redis_config_path
)
redis_creator=tfra.dynamic_embedding.RedisTableCreator(redis_config)

self.user_embeddings = tfra.dynamic_embedding.get_variable(
            name="user_dynamic_embeddings",
            dim=self.embedding_size,
            initializer=tf.keras.initializers.RandomNormal(-1, 1),
            checkpoint=False,
            KVCreator= redis_creator
            )
```

### CODE REVIEW

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_connection_pool.hpp [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+      LOG(ERROR) << "RedisHandler--other " << err.what();
+      return nullptr;
+    } catch (...) {
+      LOG(ERROR) << "RedisHandler--other crash";
+      return nullptr;
+    }
+    return nullptr;
+  }
+
+  virtual void Conn() override {
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_connection_pool.hpp:164]: virtual?

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_connection_pool.hpp [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+        throw(std::runtime_error("Exit without any Redis connection."));
+      }
+    }
+  }
+
+  static std::shared_ptr<RedisWrapper<RedisInstance, K, V>> get_instance() {
+    /* for the Meyer's Singleton mode.
+      When make Class constructor private, it will be not allow using
+      make_shared to init the Class. It' not safe, but having no choice. */
+    // static std::shared_ptr<RedisWrapper<RedisInstance, K, V>> instance_ptr =
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_connection_pool.hpp:184]: remove used code.

**Comment by MoFHeKa** [comment]
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Frecommenders-addons%2Fpull%2F117) for more info**.

@googlebot I consent.

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/json.cc [hunk 3]
```diff
@@ -0,0 +1,953 @@
+/* vim: set et ts=3 sw=3 sts=3 ft=c:
+ *
+ * Copyright (C) 2012, 2013, 2014 James McLaughlin et al.  All rights reserved.
```
Comment (Mr-Nineteen) [inline tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/json.cc:3]: 2021?

Comment (MoFHeKa) [reply tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/json.cc:3]: This is someone else's copyright notice
